plugins/inputs/directory_monitor/directory_monitor.go

```diff
@@ -0,0 +1,468 @@
+## This time can generally be low but if you choose to have a very large file written to the directory and it's potentially slow,
+## set this higher so that the plugin will wait until the file is fully copied to the directory.
+# directory_duration_threshold = "50ms"
+#
+## Character encoding to use when interpreting the file contents. Invalid
+## characters are replaced using the unicode replacement character. Defaults to utf-8.
+##   ex: character_encoding = "utf-8"
+##       character_encoding = "utf-16le"
+##       character_encoding = "utf-16be"
+# character_encoding = "utf-8"
+#
+## A list of the only file names to monitor, if necessary. Supports regex. If left blank, all files are ingested.
+# files_to_monitor = [".*.csv"]
+#
+## A list of files to ignore, if necessary. Supports regex.
+# files_to_ignore = [".DS_Store"]
+#
+## Maximum lines of the file to process that have not yet be written by the
+## output. For best throughput set based on the number of metrics and the size of the output's metric_batch_size.
+## Warning: setting this number too high can lead to a high number of dropped metrics.
+# max_buffered_metrics = 1000
+#
+## The maximum amount of files to process at once. A very high number can lead to bigger memory use and potential file system errors.
+# max_concurrent_files = 3000
+#
+## The dataformat to be read from the files.
+## Each data format has its own unique set of configuration options, read
+## more about them here:
+## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+## NOTE: We do not currently support JSON file streaming and thus JSON files will be fully loaded into memory when they are processed.
+data_format = "influx"
+`
+
+var (
+	defaultFilesToMonitor             = []string{}
+	defaultFilesToIgnore              = []string{}
+	defaultMaxBufferedMetrics         = 1000
+	defaultMaxConcurrentFiles         = 3000
+	defaultMonitorInterval            = internal.Duration{Duration: 50 * time.Millisecond}
+	defaultDirectoryDurationThreshold = internal.Duration{Duration: 50 * time.Millisecond}
+)
+
+type empty struct{}
+type semaphore chan empty
+
+type DirectoryMonitor struct {
+	Directory         string `toml:"directory"`
+	FinishedDirectory string `toml:"finished_directory"`
+	UseErrorDirectory bool   `toml:"use_error_directory"`
+	ErrorDirectory    string `toml:"error_directory"`
+
+	CharacterEncoding          string            `toml:"character_encoding"`
+	FilesToMonitor             []string          `toml:"files_to_monitor"`
+	FilesToIgnore              []string          `toml:"files_to_ignore"`
+	MaxBufferedMetrics         int               `toml:"max_buffered_metrics"`
+	MonitorInterval            internal.Duration `toml:"monitor_interval"`
+	DirectoryDurationThreshold internal.Duration `toml:"directory_duration_threshold"`
+	MaxConcurrentFiles         int               `toml:"max_concurrent_files"`
+
+	filesInUse          cmap.ConcurrentMap
```

Comment:
Would prefer using sync.Map instead of importing a new library. Any reason we can't do that? We can also potentially get rid of the shared map by using a channel to communicate instead, and getting rid of the parallelism.